### ISSUE TITLE
ZeroDivisionError: polynomial division when solving certain recurrences

### ISSUE THREAD

**Issue by ethankward** [sympy/sympy#17990]
```
>>> from sympy import *
>>> f = Function('f')
>>> x = Symbol('x')
>>> r = -10*f(x) + 4*f(x + 1) + 6*f(x + 2) + 46*f(x + 3)
>>> rsolve(r, f(x))
Traceback (most recent call last):
  File "test.py", line 148, in <module>
    print(rsolve(r, f(x)))
  File "/home/eward/se/sympy/solvers/recurr.py", line 792, in rsolve
    result = rsolve_hyper(coeffs, -i_part, n, symbols=True)
  File "/home/eward/se/sympy/solvers/recurr.py", line 623, in rsolve_hyper
    (C, s) = rsolve_poly([polys[i]*z**i for i in range(r + 1)], 0, n, symbols=True)
  File "/home/eward/se/sympy/solvers/recurr.py", line 165, in rsolve_poly
    predicate=lambda r: r >= 0).keys())
  File "/home/eward/se/sympy/polys/polyroots.py", line 974, in roots
    coeff, f = preprocess_roots(f)
  File "/home/eward/se/sympy/polys/polyroots.py", line 723, in preprocess_roots
    poly = poly.primitive()[1]
  File "/home/eward/se/sympy/polys/polytools.py", line 2923, in primitive
    cont, result = f.rep.primitive()
  File "/home/eward/se/sympy/polys/polyclasses.py", line 720, in primitive
    cont, F = dmp_ground_primitive(f.rep, f.lev, f.dom)
  File "/home/eward/se/sympy/polys/densetools.py", line 716, in dmp_ground_primitive
    return dup_primitive(f, K)
  File "/home/eward/se/sympy/polys/densetools.py", line 690, in dup_primitive
    return cont, dup_quo_ground(f, cont, K)
  File "/home/eward/se/sympy/polys/densearith.py", line 318, in dup_quo_ground
    raise ZeroDivisionError('polynomial division')
ZeroDivisionError: polynomial division
```

More examples: `f(x) + 38*f(x + 1) + 36*f(x + 2) + 41*f(x + 3)`, `-43*f(x) + 16*f(x + 1) - 36*f(x + 2) + 2*f(x + 3)`.

**Comment by jksuom** [comment]
```
f = Poly(-117968192370600*18**(1/3)/(217603955769048*(24201 + 253*sqrt(9165))**(1/3) + 2273005839412*sqrt(9165)*(24201 + 253*sqrt(9165))**(1/3)) - 15720318185*2**(2/3)*3**(1/3)*(24201 + 253*sqrt(9165))**(2/3)/(217603955769048*(24201 + 253*sqrt(9165))**(1/3) + 2273005839412*sqrt(9165)*(24201 + 253*sqrt(9165))**(1/3)) + 15720318185*12**(1/3)*(24201 + 253*sqrt(9165))**(2/3)/(217603955769048*(24201 + 253*sqrt(9165))**(1/3) + 2273005839412*sqrt(9165)*(24201 + 253*sqrt(9165))**(1/3)) + 117968192370600*2**(1/3)*3**(2/3)/(217603955769048*(24201 + 253*sqrt(9165))**(1/3) + 2273005839412*sqrt(9165)*(24201 + 253*sqrt(9165))**(1/3)), _x, domain='EX')
```
It seems that `preprocess_roots` cannot handle this complicated expression. Actually, `roots` should take care of constant polynomials before passing to polynomials of higher degree.

**Comment by jksuom** [comment]
Actually the expression is zero. It is the sum of the parameters `polys[i]*z**i` of `rsolve_poly` because they are constants, and the sum is zero because `z` is a root of the polynomial with (constant) coefficients `polys[i]`. I think that `rsolve_hyper` could have found the constant solution `C = C0` without calling `rsolve_poly`.

**Comment by ehren** [comment]
In the current main development branch this computes an incorrect 2nd order solution (see fix in #20426):

```
>>> rsolve(-10*f(x) + 4*f(x + 1) + 6*f(x + 2) + 46*f(x + 3), f(x))
C0*((86*18**(1/3)/69 + (-12 + (-1 + sqrt(3)*I)*(290412 + 3036*sqrt(9165))**(1/3))*(1 - sqrt(3)*I)*(24201 + 253*sqrt(9165))**(1/3)/276)/((1 - sqrt(3)*I)*(24201 + 253*sqrt(9165))**(1/3)))**x + C1*((86*18**(1/3)/69 + (-12 + (-1 - sqrt(3)*I)*(290412 + 3036*sqrt(9165))**(1/3))*(1 + sqrt(3)*I)*(24201 + 253*sqrt(9165))**(1/3)/276)/((1 + sqrt(3)*I)*(24201 + 253*sqrt(9165))**(1/3)))**x
```

Something changed in Poly.primitive to bypass the ZeroDivisionError but the underlying cause of it still exists (see issue #20427). (I've been sitting on these changes since a while ago when Poly.primitive still raised an error).

**Comment by ehren** [comment]
>I think that rsolve_hyper could have found the constant solution C = C0 without calling rsolve_poly.

the second commit in #20426 attempts to address this comment (the test for a constant solution has to rule out higher order polynomial solutions). That test is based on the following logic which I think is correct but doesn't result in a huge speedup in the newly added test_issue_17990_rsolve_poly (so is probably not worth complicating rsolve_poly for):

```
diff --git a/sympy/solvers/recurr.py b/sympy/solvers/recurr.py
index b2ae81b8d1..334dc03d89 100644
--- a/sympy/solvers/recurr.py
+++ b/sympy/solvers/recurr.py
@@ -47,6 +47,7 @@
     [3] a hypergeometric function  -> rsolve_hyper
 """
 from collections import defaultdict
+from itertools import islice
 
 from sympy.core.singleton import S
 from sympy.core.numbers import Rational, I
@@ -128,6 +129,8 @@ def rsolve_poly(coeffs, f, n, **hints):
     r = len(coeffs) - 1
 
     coeffs = [Poly(coeff, n) for coeff in coeffs]
+    constant_coeffs = homogeneous and max(c.degree() for c in coeffs) == 0
+    max_zero_index = -1
 
     polys = [Poly(0, n)]*(r + 1)
     terms = [(S.Zero, S.NegativeInfinity)]*(r + 1)
@@ -138,8 +141,12 @@ def rsolve_poly(coeffs, f, n, **hints):
 
         if not (polys[i].is_zero or (
                 polys[i].domain is EX and polys[i].as_expr().is_zero)):
+            if constant_coeffs:
+                break
             (exp,), coeff = polys[i].LT()
             terms[i] = (coeff, exp)
+        elif constant_coeffs:
+            max_zero_index = i
 
     d = b = terms[0][1]
 
@@ -150,6 +157,13 @@ def rsolve_poly(coeffs, f, n, **hints):
         if terms[i][1] - i > b:
             b = terms[i][1] - i
 
+    if constant_coeffs:
+        C = list(islice(numbered_symbols('C'), max_zero_index + 1))
+        result = Add(*[c*n**i for i, c in enumerate(C)])
+        if hints.get('symbols', False):
+            return result, C
+        return result
+
     d, b = int(d), int(b)
 
     x = Dummy('x')

```

The second commit in #20427 does result in a decent speed improvement however (around 5 seconds vs 24 seconds for the main test_issue_17990)
